File: cmd/videre/ui.go
package main

import (
	"fmt"
	"strconv"
	"time"
	"unicode/utf8"

	"github.com/gdamore/tcell/v2"
)

var syntaxStyleLUT = map[uint8]tcell.Style{
	hlNormal:      tcell.StyleDefault.Foreground(tcell.ColorWhite),
	hlComment:     tcell.StyleDefault.Foreground(tcell.ColorGreen),
	hlKeyword1:    tcell.StyleDefault.Foreground(tcell.ColorYellow),
	hlKeyword2:    tcell.StyleDefault.Foreground(tcell.ColorAqua),
	hlString:      tcell.StyleDefault.Foreground(tcell.ColorDarkMagenta),
	hlNumber:      tcell.StyleDefault.Foreground(tcell.ColorRed),
	hlMatch:       tcell.StyleDefault.Foreground(tcell.ColorWhite).Background(tcell.ColorNavy),
	hlMatchCursor: tcell.StyleDefault.Foreground(tcell.ColorBlack).Background(tcell.ColorYellow),
}

func runeDisplayWidth(r rune) int {
	if r < 0x20 || (r >= 0x7F && r < 0xA0) {
		return 0
	}
	// Combining characters, zero-width joiners, variation selectors
	if (r >= 0x0300 && r <= 0x036F) || (r >= 0x1AB0 && r <= 0x1AFF) ||
		(r >= 0x1DC0 && r <= 0x1DFF) || (r >= 0x20D0 && r <= 0x20FF) ||
		(r >= 0xFE20 && r <= 0xFE2F) || r == 0x200D ||
		(r >= 0xFE00 && r <= 0xFE0F) {
		return 0
	}
	// East Asian Fullwidth/Wide characters
	if (r >= 0x1100 && r <= 0x115F) || (r >= 0x2329 && r <= 0x232A) ||
		(r >= 0x2E80 && r <= 0xA4CF) || (r >= 0xAC00 && r <= 0xD7A3) ||
		(r >= 0xF900 && r <= 0xFAFF) || (r >= 0xFE10 && r <= 0xFE19) ||
		(r >= 0xFE30 && r <= 0xFE6F) || (r >= 0xFF00 && r <= 0xFF60) ||
		(r >= 0xFFE0 && r <= 0xFFE6) {
		return 2
	}
	// Emoji ranges (approximate)
	if (r >= 0x1F300 && r <= 0x1F64F) || (r >= 0x1F680 && r <= 0x1F6FF) ||
		(r >= 0x1F900 && r <= 0x1F9FF) || (r >= 0x2600 && r <= 0x27BF) {
		return 2
	}
	return 1
}

func drawRows() {
	g := gutterWidth()
	gcols := 0
	if g > 0 {
		gcols = g + 1
	}
	textCols := E.screenCols - gcols
	if textCols < 1 {
		textCols = 1
	}
	hasSelection := E.mode == modeVisual || E.mode == modeVisualLine
	lineSelection := E.mode == modeVisualLine
	sy, ey, sx, ex := 0, 0, 0, 0
	if hasSelection {
		sy, ey, sx, ex = E.selSY, E.cy, E.selSX, E.cx
		if sy > ey || (sy == ey && sx > ex) {
			sy, ey = ey, sy
			sx, ex = ex, sx
		}
	}

	for y := 0; y < E.screenRows; y++ {
		fr := y + E.rowoff
		if fr >= len(E.rows) {
			if len(E.rows) == 0 && y >= E.screenRows/3 && y < E.screenRows/3+len(welcomeLines) {
				E.Screen.SetContent(0, y, '~', nil, tcell.StyleDefault.Foreground(tcell.ColorDimGray))
				msg := welcomeLines[y-E.screenRows/3]
				padding := (textCols - len(msg)) / 2
				for i, r := range msg {
					E.Screen.SetContent(gcols+padding+i, y, r, nil, tcell.StyleDefault)
				}
			} else {
				E.Screen.SetContent(0, y, '~', nil, tcell.StyleDefault.Foreground(tcell.ColorDimGray))
			}
			continue
		}

		// Draw gutter
		if g > 0 {
			lineNum := strconv.Itoa(fr + 1)
			for i := 0; i < g-len(lineNum); i++ {
				E.Screen.SetContent(i, y, ' ', nil, tcell.StyleDefault.Foreground(tcell.ColorDimGray))
			}
			for i, r := range lineNum {
				E.Screen.SetContent(g-len(lineNum)+i, y, r, nil, tcell.StyleDefault.Foreground(tcell.ColorDimGray))
			}
			E.Screen.SetContent(g, y, ' ', nil, tcell.StyleDefault)
		}

		rowData := E.rows[fr]
		updateSyntax(rowData, false)
		line := rowData.s
		start := utf8SnapBoundary(line, E.coloff)
		
		visible := line[start:]
		hl := rowData.hl[start:]
		rowInSelection := hasSelection && fr >= sy && fr <= ey
		
		col := 0
		for i := 0; i < len(visible) && col < textCols; {
			r, n := utf8.DecodeRune(visible[i:])
			if n == 0 { break }
			
			h := hl[i]
			style := syntaxStyleLUT[h]
			
			// Selection logic
			sel := false
			if rowInSelection {
				x := i + start
				if lineSelection {
					sel = true
				} else {
					if sy == ey {
						sel = x >= sx && x <= ex
					} else if fr == sy {
						sel = x >= sx
					} else if fr == ey {
						sel = x <= ex
					} else {
						sel = true
					}
				}
			}
			if sel {
				style = style.Background(tcell.ColorDimGray)
			}

			if r == '\t' {
				tabW := 8 - ((gcols + col) % 8)
				for j := 0; j < tabW && col < textCols; j++ {
					E.Screen.SetContent(gcols+col, y, ' ', nil, style)
					col++
				}
			} else {
				E.Screen.SetContent(gcols+col, y, r, nil, style)
				col++
			}
			i += n
		}
	}
}

func drawStatusBar() {
	style := tcell.StyleDefault.Background(tcell.ColorLightGray).Foreground(tcell.ColorDarkSlateGray)
	left := " [No Name]"
	if E.filename != "" {
		left = " " + E.filename
	}
	if E.dirty {
		left += " [+]"
	}
	if E.gitStatus != "" {
		left += " [" + E.gitStatus + "]"
	}

	pos := "All"
	if len(E.rows) > 0 {
		if E.rowoff == 0 {
			pos = "Top"
		} else if E.rowoff+E.screenRows >= len(E.rows) {
			pos = "Bot"
		} else {
			pos = strconv.Itoa((E.rowoff*100)/max(1, len(E.rows)-E.screenRows)) + "%"
		}
	}

	rx := 0
	if E.cy >= 0 && E.cy < len(E.rows) {
		row := E.rows[E.cy].s
		for i := 0; i < E.cx && i < len(row); {
			if row[i] == '\t' {
				rx += 8 - (rx % 8)
				i++
				continue
			}
			r, n := utf8.DecodeRune(row[i:])
			rx += runeDisplayWidth(r)
			i += n
		}
	}
	
	loc := fmt.Sprintf("%d,%d-%d", E.cy+1, E.cx+1, rx+1)
	right := fmt.Sprintf(" %-14s %s ", loc, pos)
	
	for i := 0; i < E.screenCols; i++ {
		E.Screen.SetContent(i, E.screenRows, ' ', nil, style)
	}
	
	for i, r := range left {
		if i >= E.screenCols-len(right) { break }
		E.Screen.SetContent(i, E.screenRows, r, nil, style)
	}
	
	for i, r := range right {
		E.Screen.SetContent(E.screenCols-len(right)+i, E.screenRows, r, nil, style)
	}
}

func drawMessageBar() {
	style := tcell.StyleDefault
	y := E.screenRows + 1
	for i := 0; i < E.screenCols; i++ {
		E.Screen.SetContent(i, y, ' ', nil, style)
	}

	if E.statusmsg != "" && time.Since(E.statusTime) < 5*time.Second {
		for i, r := range E.statusmsg {
			if i >= E.screenCols { break }
			E.Screen.SetContent(i, y, r, nil, style)
		}
		return
	}

	var modeStr string
	switch E.mode {
	case modeInsert:
		modeStr = "-- INSERT --"
	case modeVisual:
		modeStr = "-- VISUAL --"
	case modeVisualLine:
		modeStr = "-- VISUAL LINE --"
	}
	for i, r := range modeStr {
		E.Screen.SetContent(i, y, r, nil, style)
	}
}

func scroll() {
	if E.rowoff < 0 { E.rowoff = 0 }
	if E.coloff < 0 { E.coloff = 0 }
	if len(E.rows) == 0 {
		E.cx, E.cy, E.preferred = 0, 0, 0
		return
	}
	if E.cy < 0 { E.cy = 0 }
	if E.cy >= len(E.rows) { E.cy = len(E.rows) - 1 }
	if E.cx < 0 { E.cx = 0 }
	if E.cx > len(E.rows[E.cy].s) { E.cx = len(E.rows[E.cy].s) }

	g := gutterWidth()
	textCols := E.screenCols - g - 1
	if textCols < 1 { textCols = 1 }
	
	if E.cy < E.rowoff { E.rowoff = E.cy }
	if E.cy >= E.rowoff+E.screenRows { E.rowoff = E.cy - E.screenRows + 1 }
	if E.cx < E.coloff { E.coloff = E.cx }
	if E.cx >= E.coloff+textCols { E.coloff = E.cx - textCols + 1 }
}

func refreshScreen() {
	updateAllSyntax(false)
	scroll()
	E.Screen.Clear()
	
	drawRows()
	drawStatusBar()
	drawMessageBar()
	drawContextMenu()
	
	// Cursor positioning
	g := gutterWidth()
	gcols := 0
	if g > 0 { gcols = g + 1 }
	
	curY := E.cy - E.rowoff
	curX := gcols
	if E.cy >= 0 && E.cy < len(E.rows) {
		line := E.rows[E.cy].s
		start := utf8SnapBoundary(line, E.coloff)
		end := E.cx
		if end > len(line) { end = len(line) }
		if end < start { end = start }
		curX += displayWidthBytes(line[start:end], gcols)
	}

	if len(E.statusmsg) > 0 && E.statusmsg[0] == ':' {
		curY = E.screenRows + 1
		curX = len(E.statusmsg)
	}
	
	E.Screen.ShowCursor(curX, curY)
	E.Screen.Show()
}

func gutterWidth() int {
	if E.filename == "" && len(E.rows) == 0 { return 0 }
	n := max(1, len(E.rows))
	w := 1
	for n >= 10 {
		n /= 10
		w++
	}
	return w
}

func displayWidthBytes(s []byte, startCol int) int {
	col := startCol
	for i := 0; i < len(s); {
		r, n := utf8.DecodeRune(s[i:])
		if n <= 0 { n = 1 }
		w := runeDisplayWidth(r)
		if r == '\t' {
			w = 8 - (col % 8)
		}
		col += w
		i += n
	}
	return col - startCol
}

var welcomeLines = []string{
	"VIDERE v0.1.0",
	"",
	"videre is open source and freely distributable",
	"https://github.com/euxaristia/videre",
	"",
	"type  :q<Enter>               to exit         ",
	"type  :wq<Enter>              save and exit   ",
	"",
	"Maintainer: euxaristia",
}

var menuItems = []string{
	" Cut       ",
	" Copy      ",
	" Paste     ",
	" Select All ",
	"----------- ",
	" Undo      ",
	" Redo      ",
}

var contextMenuW int

func initContextMenuMetrics() {
	w := 1
	for _, item := range menuItems {
		if len(item) > w {
			w = len(item)
		}
	}
	contextMenuW = w
}

func drawContextMenu() {
	// Implementation for tcell menu if needed
}

func updateWindowSize() {
	w, h := E.Screen.Size()
	E.screenCols = w
	E.screenRows = h - 2
}

func byteIndexFromDisplayCol(s []byte, target int, colStart int) int {
	if target <= colStart {
		return 0
	}
	i := 0
	col := colStart
	for i < len(s) {
		r, n := utf8.DecodeRune(s[i:])
		if n <= 0 { n = 1 }
		w := runeDisplayWidth(r)
		if r == '\t' {
			tabW := 8 - (col % 8)
			if tabW == 0 { tabW = 8 }
			w = tabW
		}
		if col+w > target { break }
		col += w
		i += n
	}
	return i
}

File: src/ui.rs
use crossterm::{
    cursor,
    queue,
    style::{Color, Print, SetBackgroundColor, SetForegroundColor},
    terminal::{self, Clear, ClearType},
};
use std::io::{stdout, Result, Write};

use crate::types::{Highlight, Mode};
use crate::editor::{Editor, utf8_snap_boundary, display_width_bytes, rune_display_width};
use crate::syntax;

pub fn enable_raw_mode() -> Result<()> {
    terminal::enable_raw_mode()
}

fn highlight_to_color(hl: Highlight) -> (Color, Color) {
    match hl {
        Highlight::Normal => (Color::White, Color::Reset),
        Highlight::Comment => (Color::Green, Color::Reset),
        Highlight::Keyword1 => (Color::Yellow, Color::Reset),
        Highlight::Keyword2 => (Color::Cyan, Color::Reset),
        Highlight::String => (Color::Magenta, Color::Reset),
        Highlight::Number => (Color::Red, Color::Reset),
        Highlight::Match => (Color::White, Color::DarkBlue),
        Highlight::MatchCursor => (Color::Black, Color::Yellow),
        Highlight::Visual => (Color::White, Color::DarkGrey),
    }
}

pub fn disable_raw_mode() -> Result<()> {
    terminal::disable_raw_mode()
}

pub fn get_terminal_size() -> Result<(usize, usize)> {
    let (cols, rows) = terminal::size()?;
    Ok((cols as usize, rows as usize))
}

pub fn refresh_screen(editor: &mut Editor) -> Result<()> {
    editor.scroll();
    syntax::update_all_syntax(&editor.filename, &editor.search_regexp, &mut editor.rows, false);
    let mut stdout = stdout();

    queue!(stdout, cursor::Hide, cursor::MoveTo(0, 0))?;

    draw_rows(editor, &mut stdout)?;
    draw_status_bar(editor, &mut stdout)?;
    draw_message_bar(editor, &mut stdout)?;

    let g = editor.gutter_width();
    let gcols = if g > 0 { g + 1 } else { 0 };
    
    let mut cur_y = (editor.cy.saturating_sub(editor.rowoff)) as u16;
    let mut cur_x = gcols as u16;

    if editor.cy < editor.rows.len() {
        let row = &editor.rows[editor.cy].s;
        let start = utf8_snap_boundary(row, editor.coloff);
        let end = editor.cx.min(row.len());
        if end > start {
            cur_x += display_width_bytes(&row[start..end], gcols) as u16;
        }
    }

    if !editor.statusmsg.is_empty() && editor.statusmsg.starts_with(':') {
        cur_y = (editor.screen_rows + 1) as u16;
        cur_x = editor.statusmsg.len() as u16;
    }

    queue!(stdout, cursor::MoveTo(cur_x, cur_y), cursor::Show)?;
    stdout.flush()?;
    Ok(())
}

const WELCOME_LINES: &[&str] = &[
    "VIDERE v0.1.0",
    "",
    "videre is open source and freely distributable",
    "https://github.com/euxaristia/videre",
    "",
    "type  :q<Enter>               to exit         ",
    "type  :wq<Enter>              save and exit   ",
    "",
    "Maintainer: euxaristia",
];

fn draw_rows(editor: &mut Editor, stdout: &mut impl Write) -> Result<()> {
    let g = editor.gutter_width();
    let gcols = if g > 0 { g + 1 } else { 0 };
    let text_cols = editor.screen_cols.saturating_sub(gcols);

    let has_selection = editor.mode == Mode::Visual || editor.mode == Mode::VisualLine;
    let line_selection = editor.mode == Mode::VisualLine;
    let (mut sy, mut ey, mut sx, mut ex) = (0, 0, 0, 0);
    if has_selection {
        sy = editor.sel_sy; ey = editor.cy;
        sx = editor.sel_sx; ex = editor.cx;
        if sy > ey || (sy == ey && sx > ex) {
            std::mem::swap(&mut sy, &mut ey);
            std::mem::swap(&mut sx, &mut ex);
        }
    }

    for y in 0..editor.screen_rows {
        let filerow = y + editor.rowoff;
        if filerow >= editor.rows.len() {
            if editor.rows.is_empty() && y >= editor.screen_rows / 3 && y < editor.screen_rows / 3 + WELCOME_LINES.len() {
                queue!(stdout, SetForegroundColor(Color::DarkGrey), Print("~"), SetForegroundColor(Color::Reset))?;
                
                let msg = WELCOME_LINES[y - editor.screen_rows / 3];
                let padding = text_cols.saturating_sub(msg.len()) / 2;
                for _ in 0..padding {
                    queue!(stdout, Print(" "))?;
                }
                let display_len = std::cmp::min(msg.len(), text_cols.saturating_sub(padding));
                queue!(stdout, Print(&msg[..display_len]))?;
            } else {
                queue!(stdout, SetForegroundColor(Color::DarkGrey), Print("~"), SetForegroundColor(Color::Reset))?;
            }
        } else {
            if g > 0 {
                let line_num = (filerow + 1).to_string();
                let padding = g.saturating_sub(line_num.len());
                queue!(stdout, SetForegroundColor(Color::DarkGrey))?;
                for _ in 0..padding {
                    queue!(stdout, Print(" "))?;
                }
                queue!(stdout, Print(line_num))?;
                queue!(stdout, Print(" "), SetForegroundColor(Color::Reset))?;
            }

            let row = &editor.rows[filerow];
            let start = utf8_snap_boundary(&row.s, editor.coloff);
            let row_in_selection = has_selection && filerow >= sy && filerow <= ey;

            let mut col = 0;
            let mut i = start;
            while i < row.s.len() && col < text_cols {
                let (r, n) = crate::editor::decode_utf8_rune(&row.s[i..]);
                if n == 0 { break; }

                let mut hl = row.hl[i];
                // Check if current search match is under cursor
                if hl == Highlight::Match
                    && let Some(re) = &editor.search_regexp
                        && let Some(m) = re.find(&row.s)
                            && filerow == editor.cy && i >= m.start() && i < m.end() && editor.cx >= m.start() && editor.cx < m.end() {
                                hl = Highlight::MatchCursor;
                            }

                let (fg, mut bg) = highlight_to_color(hl);

                if row_in_selection {
                    let x = i;
                    let sel = if line_selection { true }
                            else if sy == ey { x >= sx && x <= ex }
                            else if filerow == sy { x >= sx }
                            else if filerow == ey { x <= ex }
                            else { true };
                    if sel { bg = Color::DarkGrey; }
                }

                queue!(stdout, SetForegroundColor(fg), SetBackgroundColor(bg))?;
                if r == '\t' {
                    let tab_w = 8 - ((gcols + col) % 8);
                    for _ in 0..tab_w {
                        if col < text_cols {
                            queue!(stdout, Print(" "))?;
                            col += 1;
                        }
                    }
                } else {
                    let w = rune_display_width(r);
                    if col + w <= text_cols {
                        queue!(stdout, Print(r))?;
                        col += w;
                    } else {
                        break;
                    }
                }
                i += n;
            }
            queue!(stdout, SetForegroundColor(Color::Reset), SetBackgroundColor(Color::Reset))?;
        }
        
        queue!(stdout, Clear(ClearType::UntilNewLine))?;
        queue!(stdout, Print("\r\n"))?;
    }
    Ok(())
}

fn draw_status_bar(editor: &Editor, stdout: &mut impl Write) -> Result<()> {
    let mut left = if editor.filename.is_empty() {
        String::from(" [No Name]")
    } else {
        format!(" {}", editor.filename)
    };
    if editor.dirty {
        left.push_str(" [+]");
    }
    if !editor.git_status.is_empty() {
        left.push_str(&format!(" [{}]", editor.git_status));
    }

    let pos = if editor.rows.is_empty() {
        String::from("All")
    } else if editor.rowoff == 0 {
        String::from("Top")
    } else if editor.rowoff + editor.screen_rows >= editor.rows.len() {
        String::from("Bot")
    } else {
        let percent = (editor.rowoff * 100) / (editor.rows.len() - editor.screen_rows).max(1);
        format!("{}%", percent)
    };

    let rx = editor.get_rx();
    let loc = format!("{}:{},{}-{}", editor.filename, editor.cy + 1, editor.cx + 1, rx + 1);
    let right = format!(" {} {} ", loc, pos);

    queue!(
        stdout,
        SetBackgroundColor(Color::Grey),
        SetForegroundColor(Color::Black)
    )?;

    let left_len = std::cmp::min(left.len(), editor.screen_cols.saturating_sub(right.len()));
    queue!(stdout, Print(&left[..left_len]))?;
    
    let padding = editor.screen_cols.saturating_sub(left_len).saturating_sub(right.len());
    for _ in 0..padding {
        queue!(stdout, Print(" "))?;
    }
    
    if editor.screen_cols >= left_len + right.len() {
        queue!(stdout, Print(right))?;
    }
    
    queue!(
        stdout,
        SetBackgroundColor(Color::Reset),
        SetForegroundColor(Color::Reset),
        Print("\r\n")
    )?;
    Ok(())
}

fn draw_message_bar(editor: &Editor, stdout: &mut impl Write) -> Result<()> {
    queue!(stdout, Clear(ClearType::UntilNewLine))?;
    
    let mut msg = editor.statusmsg.clone();
    if msg.is_empty() {
        msg = match editor.mode {
            Mode::Insert => String::from("-- INSERT --"),
            Mode::Visual => String::from("-- VISUAL --"),
            Mode::VisualLine => String::from("-- VISUAL LINE --"),
            Mode::Replace => String::from("-- REPLACE --"),
            _ => String::new(),
        };
    }
    
    let len = std::cmp::min(msg.len(), editor.screen_cols);
    queue!(stdout, Print(&msg[..len]))?;
    Ok(())
}
